README.md
<a href = "https://appsilon.com/careers/" target="_blank"><img src="http://d2v95fjda94ghc.cloudfront.net/hiring.png" alt="We are hiring!"/></a>

<img src="man/figures/hexsticker.png" align="right" alt="" width="130" />

# shiny.layouts

<!-- badges: start -->

![R-CMD-check](https://github.com/Appsilon/shiny.layouts/workflows/R-CMD-check/badge.svg)
[![codecov](https://codecov.io/gh/Appsilon/shiny.layouts/branch/master/graph/badge.svg)](https://codecov.io/gh/Appsilon/shiny.layouts)

[![cranlogs](https://cranlogs.r-pkg.org/badges/shiny.layouts)](https://CRAN.R-project.org/package=shiny.layouts)
[![total](https://cranlogs.r-pkg.org/badges/grand-total/shiny.layouts)](https://CRAN.R-project.org/package=shiny.layouts)
<!-- badges: end -->

**CSS Grid and pre-defined useful layouts for Shiny**

With this library it is easy to define custom layouts in Shiny apps using the powerful
**[CSS grid](https://css-tricks.com/snippets/css/complete-guide-grid/)**.
It also provides a number of ready to use layouts like sidebar layout, vertical layout etc.

It works great together with [shiny.semantic](https://github.com/Appsilon/shiny.semantic), but is compatible with any UI framework.

## How to install?

You can install a stable `shiny.layouts` release from CRAN repository:

``` r
install.packages("shiny.layouts")
```

and the latest version with `remotes`:

``` r
remotes::install_github("Appsilon/shiny.layouts@develop")
```

(`main` branch contains the stable version. Use `develop` branch for
latest features)


## How to contribute?

We welcome contributions of all types\!

We encourage typo corrections, bug reports, bug fixes and feature
requests. Feedback on the clarity of the documentation and examples is
especially valuable.

If you want to contribute to this project please submit a regular PR,
once you’re done with new feature or bug fix.<br>

**Changes in documentation**

Documentation is rendered with `pkgdown`. Just run
`pkgdown::build_site()` after making changes.

**Changes in vignettes**

If you modified or added a new vignette, please rebuild them with
`devtools::build_vignettes()`.

## Appsilon Data Science

<img src="https://avatars0.githubusercontent.com/u/6096772" align="right" alt="" width="6%" />

Appsilon is the **Full Service Certified RStudio Partner**. Learn more
at [appsilon.com](https://appsilon.com).

Get in touch [dev@appsilon.com](dev@appsilon.com)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
## [0.1.0]

*2021-02-11*

### Added

- layouts and grid extracted from shiny.semantic.


tests/testthat.R
library(testthat)
library(shiny.layouts)

test_check("shiny.layouts")


tests/testthat/test_grid.R
context("grid")
library(htmltools)

test_that("test data_frame_to_css_grid_template_areas", {
  areas_dataframe <- rbind(
     c("header", "header", "header"),
     c("menu",   "main",   "right1"),
     c("menu",   "main",   "right2")
  )
  expect_equal(
    data_frame_to_css_grid_template_areas(areas_dataframe),
    "'header header header' 'menu main right1' 'menu main right2'"
  )

  areas_dataframe <- rbind(
    c("a", "b"),
    c("c", "d")
  )
  expect_equal(
    data_frame_to_css_grid_template_areas(areas_dataframe),
    "'a b' 'c d'"
  )
})


test_that("test grid_container_css", {
  expected_result <- paste(
    "display: grid;",
    "height: 100%;",
    "grid-template-rows: 50% 50%;",
    "grid-template-columns: 100px 2fr 1fr;",
    "grid-template-areas: 'a a a' 'b b b';",
    "{{ custom_style_grid_container }}"
  )
  expect_equal(
    grid_container_css("'a a a' 'b b b'", c("50%", "50%"), c("100px", "2fr", "1fr")),
    expected_result
  )
})


test_that("test list_of_area_tags", {
  expected_result <- list(
    HTML('<div id="{{ grid_id }}-header" style="grid-area: header; {{ header_custom_css }}">{{ header }}</div>'),
    HTML('<div id="{{ grid_id }}-main" style="grid-area: main; {{ main_custom_css }}">{{ main }}</div>'),
    HTML('<div id="{{ grid_id }}-footer" style="grid-area: footer; {{ footer_custom_css }}">{{ footer }}</div>')
  )
  expect_equal(list_of_area_tags(c("header", "main", "footer")), expected_result)
  expect_equal(list_of_area_tags(c()), list())
})


remove_whitespace <- function (string) gsub('[\n ]', '', as.character(string))

test_that("test grid_template", {
  myGrid <- grid_template(default = list(
   areas = rbind(
     c("header", "header", "header"),
     c("menu",   "main",   "right1"),
     c("menu",   "main",   "right2")
   ),
   rows_height = c("50px", "auto", "100px"),
   cols_width = c("100px", "2fr", "1fr")
  ))

  expected_template <- HTML('
    <style>
      #{{ grid_id }} {
        display: grid;
        height: 100%;
        grid-template-rows: 50px auto 100px;
        grid-template-columns: 100px 2fr 1fr;
        grid-template-areas: \'header header header\' \'menu main right1\' \'menu main right2\';
        {{ custom_style_grid_container }}
      }
    </style>
    <div id="{{ grid_id }}">
      <div id="{{ grid_id }}-header" style="grid-area: header; {{ header_custom_css }}">{{ header }}</div>
      <div id="{{ grid_id }}-menu" style="grid-area: menu; {{ menu_custom_css }}">{{ menu }}</div>
      <div id="{{ grid_id }}-main" style="grid-area: main; {{ main_custom_css }}">{{ main }}</div>
      <div id="{{ grid_id }}-right1" style="grid-area: right1; {{ right1_custom_css }}">{{ right1 }}</div>
      <div id="{{ grid_id }}-right2" style="grid-area: right2; {{ right2_custom_css }}">{{ right2 }}</div>
    </div>
  ')

  expected_area_names <- c("header", "menu", "main", "right1", "right2")

  expect_equal(remove_whitespace(myGrid$template), remove_whitespace(expected_template))
  expect_equal(myGrid$area_names, expected_area_names)
})


test_that("test grid_template with mobile version", {
  myGrid <- grid_template(
    default = list(
      areas = rbind(
        c("header", "header", "header"),
        c("menu",   "main",   "right1"),
        c("menu",   "main",   "right2")
      ),
      rows_height = c("50px", "auto", "100px"),
      cols_width = c("100px", "2fr", "1fr")
    ),
    mobile = list(
      areas = rbind(
        "header",
        "menu",
        "main",
        "right1",
        "right2"
      ),
      rows_height = c("50px", "50px", "auto", "100px", "100px"),
      cols_width = c("100%")
    )
  )

  expected_template <- HTML('
    <style>
      #{{ grid_id }} {
        display: grid;
        height: 100%;
        grid-template-rows: 50px auto 100px;
        grid-template-columns: 100px 2fr 1fr;
        grid-template-areas: \'header header header\' \'menu main right1\' \'menu main right2\';
        {{ custom_style_grid_container }}
      }
    </style>
    <style>
      @media screen and (max-width: 768px) {
        #{{ grid_id }} {
          display: grid;
          height: 100%;
          grid-template-rows: 50px 50px auto 100px 100px;
          grid-template-columns: 100%;
          grid-template-areas: \'header\' \'menu\' \'main\' \'right1\' \'right2\';
          {{ custom_style_grid_container }}
        }
      }
    </style>
    <div id="{{ grid_id }}">
      <div id="{{ grid_id }}-header" style="grid-area: header; {{ header_custom_css }}">{{ header }}</div>
      <div id="{{ grid_id }}-menu" style="grid-area: menu; {{ menu_custom_css }}">{{ menu }}</div>
      <div id="{{ grid_id }}-main" style="grid-area: main; {{ main_custom_css }}">{{ main }}</div>
      <div id="{{ grid_id }}-right1" style="grid-area: right1; {{ right1_custom_css }}">{{ right1 }}</div>
      <div id="{{ grid_id }}-right2" style="grid-area: right2; {{ right2_custom_css }}">{{ right2 }}</div>
    </div>
  ')

  expected_area_names <- c("header", "menu", "main", "right1", "right2")

  expect_equal(remove_whitespace(myGrid$template), remove_whitespace(expected_template))
  expect_equal(myGrid$area_names, expected_area_names)
})


test_that("test grid", {
  simpleGrid <- grid_template(
    default = list(
      areas = rbind(c("a","b")),
      rows_height = "100%",
      cols_width = "50%", "50%"
    )
  )

  gridHTML <- grid(simpleGrid,
                   id = "test_grid",
                   container_style = "",
                   area_styles = list(),
                   a = shiny::tags$div("hello"),
                   b = shiny::tags$div("world"))

  expected_html <- HTML('
    <style>
      #test_grid {
        display: grid;
        height: 100%;
        grid-template-rows: 100%;
        grid-template-columns: 50%;
        grid-template-areas: \'a b\';
      }
    </style>
    <div id="test_grid">
      <div id="test_grid-a" style="grid-area: a; "><div>hello</div></div>
      <div id="test_grid-b" style="grid-area: b; "><div>world</div></div>
    </div>
  ')

  expect_equal(remove_whitespace(gridHTML), remove_whitespace(expected_html))

  # Display mode
  gridHTML <- grid(simpleGrid,
                   id = "test_grid",
                   container_style = "",
                   area_styles = list(),
                   display_mode = TRUE,
                   a = shiny::tags$div("hello"),
                   b = shiny::tags$div("world"))

  expected_html <- HTML('
    <style>
      #test_grid {
        display: grid;
        height: 100%;
        grid-template-rows: 100%;
        grid-template-columns: 50%;
        grid-template-areas: \'a b\';
      }
    </style>
    <div id="test_grid">
      <div id="test_grid-a" style="grid-area: a; ">a</div>
      <div id="test_grid-b" style="grid-area: b; ">b</div>
    </div>
  ')

  expect_equal(remove_whitespace(gridHTML), remove_whitespace(expected_html))
})


tests/testthat/test_layouts.R
context("layouts")
library(htmltools)

test_that("test sidebar_panel", {
  # type
  expect_is(sidebar_panel(), "list")
  # simple input
  si_str <- as.character(sidebar_panel(h1("abc"), p("qqq"))$children)
  expect_true(any(grepl("<h1>abc</h1>",
                        si_str, fixed = TRUE)))
  expect_true(any(grepl("qqq",
                        si_str, fixed = TRUE)))
})

test_that("test main_panel", {
  # type
  expect_is(main_panel(), "list")
  # simple input
  si_str <- as.character(main_panel(h1("abc"), p("qqq"))$children)
  expect_true(any(grepl("<h1>abc</h1>",
                        si_str, fixed = TRUE)))
  expect_true(any(grepl("qqq",
                        si_str, fixed = TRUE)))
})

test_that("test sidebar_layout", {
  # type
  expect_error(sidebar_layout(), "argument \"sidebar_panel\" is missing")
  expect_error(sidebar_layout(sidebar_panel()),
               "argument \"main_panel\" is missing")
  # empty input
  expect_is(sidebar_layout(sidebar_panel(),main_panel()), "list")
})

test_that("test split_layout", {
  # empty input gives error
  expect_error(sidebar_layout())
  # simple input
  si_str <- as.character(split_layout(p("a")))
  expect_true(any(grepl("<p>a</p>", si_str, fixed = TRUE)))
  expect_true(any(grepl("col1", si_str, fixed = TRUE)))
  expect_false(any(grepl("col2", si_str, fixed = TRUE)))

  # width of columns (number)
  si_str <- as.character(split_layout(
    cell_widths = 300,
    cell_args = "padding: 6px;",
    p("p1")
  ))
  expect_true(any(grepl("grid-template-columns: 300", si_str, fixed = TRUE)))
  # width of columns (vector)
  si_str <- as.character(
    split_layout(cell_widths = c("25%", "75%"), p("p1"), p("p2"))
  )
  expect_true(any(grepl("grid-template-columns: 25% 75%", si_str, fixed = TRUE)))
  si_str <- as.character(
    split_layout(p("p1"), p("p2"), style = "background:red;")
  )
  expect_true(any(grepl("background:red;", si_str, fixed = TRUE)))
})

test_that("test splitLayout", {
  # test equivalence
  set.seed(123) # For generating the same random CSS id selector for grid.
  layout1 <- split_layout(p("a"))
  set.seed(123)
  layout2 <- split_layout(p("a"))
  expect_equal(layout1, layout2)

  set.seed(123)
  layout1 <- split_layout(p("a"), cell_widths = c("25%", "75%"))
  set.seed(123)
  layout2 <- splitLayout(p("a"), cellWidths = c("25%", "75%"))
  expect_equal(layout1, layout2)
})

test_that("test vertical_layout", {
  # empty input gives error
  expect_error(vertical_layout())
  # simple input
  si_str <- as.character(vertical_layout(h1("H"), p("a")))
  expect_true(any(grepl("<h1>H</h1>", si_str, fixed = TRUE)))
  expect_true(any(grepl("<p>a</p>", si_str, fixed = TRUE)))
  expect_true(any(grepl("row1", si_str, fixed = TRUE)))
  expect_false(any(grepl("row22", si_str, fixed = TRUE)))

  # check parameters passed to cell
  si_str <- as.character(vertical_layout(
    cell_args = "padding: 6px; background: red;",
    p("p1")
  ))
  expect_true(any(grepl("background: red", si_str, fixed = TRUE)))

  # test row_heights param
  si_str <- as.character(vertical_layout(h1("H"), p("a"), rows_heights = "20px"))
  expect_true(any(grepl("grid-template-rows: 20px 20px;", si_str, fixed = TRUE)))
  si_str <- as.character(vertical_layout(h1("H"), p("a"),
                                         rows_heights = c("20px", "40px")))
  expect_true(any(grepl("grid-template-rows: 20px 40px;", si_str, fixed = TRUE)))
})


test_that("test verticalLayout", {
  set.seed(123) # For generating the same random CSS id selector for grid.
  v1 <- vertical_layout(p("a"), adjusted_to_page = F)
  set.seed(123)
  v2 <- verticalLayout(p("a"))
  expect_equal(v1,v2)
})

test_that("test flow_layout", {
  actual <- as.character(flow_layout(
    cell_args = list(class = "cell"),
    min_cell_width = "30%",
    column_gap = "15px",
    row_gap = 10,
    shiny::tags$p("a"),
    shiny::tags$p("b")
  ))
  has <- function(expected) {
    any(grepl(expected, actual, fixed = TRUE))
  }
  expect_true(has("display: grid"))
  expect_true(has("align-self: start"))
  expect_true(has('class="cell"'))
  expect_true(has("grid-template-columns: repeat(auto-fill, minmax(30%, 1fr));"))
  expect_true(has("column-gap: 15px"))
  expect_true(has("row-gap: 10px"))
  expect_true(has("<p>a</p>"))
  expect_true(has("<p>b</p>"))
})

test_that("test flowLayout", {
  actual <- as.character(flowLayout(
    cellArgs = list(class = "cell"),
    shiny::tags$p("a")
  ))
  expected <- as.character(flow_layout(
    cell_args = list(class = "cell"),
    shiny::tags$p("a")
  ))
  expect_equal(actual, expected)
})


examples/grid/app.R
library(shiny)
library(shiny.layouts)

# Here we define the grid_template that we will pass to our grid
myGrid <- grid_template(
  default = list(
    # Here we define the data.frame describing our layout
    # The easiest way is to use rbind so that the layout can be 'visualized' in code
    areas = rbind(
      c("header", "header", "header"),
      c("menu",   "main",   "right1"),
      c("menu",   "main",   "right2")
    ),
    # Then we define the dimensions of the different elements of the layout
    # We can use any valid css units to make the layout behave exactly as desired
    rows_height = c("50px", "auto", "100px"),
    cols_width = c("100px", "2fr", "1fr")
  ),
  # This is optional, but we can define a specific layout for mobile (screen width below 768px)
  mobile = list(
    areas = rbind(
      c("header", "header", "header"),
      c("menu",   "main",   "right1"),
      c("menu",   "main",   "right2")
    ),
    rows_height = c("50px", "2fr", "1fr"), # Notice how we changed the rows heights here
    cols_width = c("100px", "2fr", "1fr")
  )
)

# We can use nested grids to precisely control the layout of all the elements
# Here we define another grid_template to use in one of the elements of the parent grid
subGrid <- grid_template(
  default = list(
    areas = rbind(
      c("top_left", "top_right"),
      c("bottom_left", "bottom_right")
    ),
    rows_height = c("50%", "50%"),
    cols_width = c("50%", "50%")
  ),
  # This is optional, but we can define a specific layout for mobile (screen width below 768px)
  mobile = list(
    areas = rbind(
      c("top_left"),
      c("top_right"),
      c("bottom_left"),
      c("bottom_right")
    ),
    rows_height = c("25%", "25%", "25%", "25%"),
    cols_width = c("100%")
  )
)

ui <- bootstrapPage(
  grid(myGrid,
       # We can define the css style of the grid using container_style
       container_style = "border: 5px solid #3d7ea6",
       # We can define the css style of each of the grid elements using area_styles
       area_styles = list(header = "border: 3px solid #5c969e",
                          menu = "border: 3px solid #5c969e",
                          main = "border: 3px solid #5c969e",
                          right1 = "border: 3px solid #5c969e",
                          right2 = "border: 3px solid #5c969e"),
       # Finally, we define the ui content we would like to have inside each element
       header = div(shiny::tags$h1("Hello CSS Grid!")),
       menu = div("menu"),
       main = grid(subGrid,
                   area_styles = list(top_left = "border: 3px solid #ffa5a5;",
                                      top_right = "border: 3px solid #ffa5a5;",
                                      bottom_left = "border: 3px solid #ffa5a5;",
                                      bottom_right = "border: 3px solid #ffa5a5;"),
                   top_left = div("main top left"),
                   top_right = div("main top right"),
                   bottom_left = div("main bottom left"),
                   bottom_right = div("main bottom right")
       ),
       right1 = div("right 1"),
       right2 = div("right 2")
  )
)

server <- function(input, output, session) {
}

shinyApp(ui = ui, server = server)


examples/layouts/sidebar_app.R
library(shiny)
library(shiny.layouts)
library(shiny.semantic)

grid_cell_style = "
  align-items: center;
  border: 1px solid #ccc;
  display: flex;
  justify-content: center;
"

ui <- function() {
  shinyUI(semanticPage(
    title = "Sidebar Layout Test",
    sidebar_layout(
      sidebar_panel(
        width = 1,
        h1("Hello shiny.semantic!"),
        sliderInput("obs1", "Select value A:", min = 0, max = 800, value = 200),
        sliderInput("obs2", "Select value B:", min = 0, max = 1000, value = 400),
        sliderInput("obs3", "Select value C:", min = 0, max = 600, value = 500),
        sliderInput("obs4", "Select value D:", min = 0, max = 700, value = 300)
      ),
      main_panel(
        width = 2,
        h2("Selected values:"),
        grid(
          grid_template = grid_template(default = list(
            areas = rbind(
              c("top_left", "top_right"),
              c("bottom_left", "bottom_right")
            ),
            cols_width = c("1fr", "1fr"),
            rows_height = c("1fr", "1fr")
          )),
          container_style = "gap: 10px;",
          area_styles = list(
            top_left = grid_cell_style,
            top_right = grid_cell_style,
            bottom_left = grid_cell_style,
            bottom_right = grid_cell_style
          ),
          top_left = h3(textOutput("obs1")),
          top_right = h3(textOutput("obs2")),
          bottom_left = h3(textOutput("obs3")),
          bottom_right = h3(textOutput("obs4")),
        ),
        p("Have a nice day!")
      ),
      min_height = "400px",
      mirrored = FALSE,
      container_style = "background-color: white;",
      area_styles = list(
        sidebar_panel = "border: 1px solid #888;",
        main_panel = "border: 1px solid #ccc;"
      )
    )
  ))
}

server <- shinyServer(function(input, output) {
  output$obs1 <- renderText({ glue::glue("Value A: {input$obs1}") })
  output$obs2 <- renderText({ glue::glue("Value B: {input$obs2}") })
  output$obs3 <- renderText({ glue::glue("Value C: {input$obs3}") })
  output$obs4 <- renderText({ glue::glue("Value D: {input$obs4}") })
})

shinyApp(ui = ui(), server = server)


R/grid.R
#' Generate CSS string representing grid template areas.
#' @param areas_dataframe data.frame of character representing grid areas
#' @return character
#' @details This is a helper function used in grid_template()
#'
#' \preformatted{
#' areas_dataframe <- rbind(
#'    c("header", "header", "header"),
#'    c("menu",   "main",   "right1"),
#'    c("menu",   "main",   "right2")
#' )
#'
#' result == "'header header header' 'menu main right1' 'menu main right2'"
#' }
#'
data_frame_to_css_grid_template_areas <- function(areas_dataframe) {
  apply(areas_dataframe, 1, function(row) paste(row, collapse = ' ')) %>%
    lapply(function(row) glue::glue("'{row}'")) %>%
    paste(collapse = ' ')
}

#' Generate template string representing CSS styles of grid container div.
#'
#' @param css_grid_template_areas character, CSS value for grid-template-areas
#' @param rows_height vector of character
#' @param cols_width vector of character
#'
#' @return character
#' @details This is a helper function used in grid_template()
#' \preformatted{
#'   grid_container_css(
#'     "'a a a' 'b b b'",
#'     c("50\%", "50\%"),
#'     c("100px", "2fr", "1fr")
#'   )
#' }
#' returns
#' \preformatted{
#'   "display: grid;
#'    height: 100\%;
#'    grid-template-rows: 50\% 50\%;
#'    grid-template-columns: 100px 2fr 1fr;
#'    grid-template-areas: 'a a a' 'b b b';
#'    {{ custom_style_grid_container }}"
#' }
#'
grid_container_css <- function(css_grid_template_areas, rows_height, cols_width) {
  grid_container_styles <- c(
    "display: grid",
    "height: 100%",
    glue::glue("grid-template-rows: {paste(rows_height, collapse = ' ')}"),
    glue::glue("grid-template-columns: {paste(cols_width, collapse = ' ')}"),
    glue::glue("grid-template-areas: {css_grid_template_areas}"),
    "{{ custom_style_grid_container }}" # To be rendered by htmltools::htmlTemplate() later
  )
  paste(grid_container_styles, collapse = "; ")
}

#' Generate list of HTML div elements representing grid areas.
#'
#' @param area_names vector with area names
#'
#' @return list of \code{shiny::tags$div}
#'
#' @details This is a helper function used in grid_template()
#' \preformatted{
#'   list_of_area_tags(c("header", "main", "footer"))
#' }
#' returns the following list:
#' \preformatted{
#'   [[1]] <div id="{{ grid_id }}-header" style="grid-area: header; {{ header_custom_css }}">{{ header }}</div>
#'   [[2]] <div id="{{ grid_id }}-main" style="grid-area: main; {{ main_custom_css }}">{{ main }}</div>
#'   [[3]] <div id="{{ grid_id }}-footer" style="grid-area: footer; {{ footer_custom_css }}">{{ footer }}</div>
#' }
#'
list_of_area_tags <- function(area_names) {
  lapply(area_names,
    function(name) {
      mustache_id <- "{{ grid_id }}"
      mustache_css <- paste0("{{ ", name, "_custom_css }}")
      mustache_area <- paste("{{", name, "}}")
      return(htmltools::HTML(
        glue::glue('<div id="{mustache_id}-{name}" style="grid-area: {name}; {mustache_css}">{mustache_area}</div>')
      ))
    }
  )
}

#' Define a template of a CSS grid
#'
#' @param default (required)
#' Template for desktop:
#' list(areas = [data.frame of character],
#'      rows_height = [vector of character],
#'      cols_width = [vector of character])
#' @param mobile (optional)
#' Template for mobile (screen width below 768px):
#' list(areas = [data.frame of character],
#'      rows_height = [vector of character],
#'      cols_width = [vector of character])
#'
#' @return
#' list(template = [character], area_names = [vector of character])
#'
#' template - contains template that can be parsed by htmlTemplate() function
#'
#' area_names - contains all unique area names used in grid definition
#'
#' @examples
#' myGrid <- grid_template(
#'   default = list(
#'     areas = rbind(
#'       c("header", "header", "header"),
#'       c("menu",   "main",   "right1"),
#'       c("menu",   "main",   "right2")
#'     ),
#'     rows_height = c("50px", "auto", "100px"),
#'     cols_width = c("100px", "2fr", "1fr")
#'   ),
#'   mobile = list(
#'     areas = rbind(
#'       "header",
#'       "menu",
#'       "main",
#'       "right1",
#'       "right2"
#'     ),
#'     rows_height = c("50px", "50px", "auto", "150px", "150px"),
#'     cols_width = c("100%")
#'   )
#' )
#' if (interactive()) display_grid(myGrid)
#' subGrid <- grid_template(default = list(
#'   areas = rbind(
#'     c("top_left", "top_right"),
#'     c("bottom_left", "bottom_right")
#'   ),
#'   rows_height = c("50%", "50%"),
#'   cols_width = c("50%", "50%")
#' ))
#'
#' if (interactive()) display_grid(subGrid)
#' @importFrom magrittr %>%
#' @export
grid_template <- function(default = NULL, mobile = NULL) {

  if (!("areas" %in% names(default))) {
    stop(paste("grid_template() default argument must contain list with `areas` definition.",
               "See documentation for examples."))
  }

  area_names <- default$areas %>% as.vector %>% unique
  area_tags <- shiny::tagList(list_of_area_tags(area_names))

  css_grid_template_areas <- data_frame_to_css_grid_template_areas(default$areas)
  css_default <- shiny::tags$style(paste(
    "#{{ grid_id }} {",
      grid_container_css(css_grid_template_areas, default$rows_height, default$cols_width),
    "}"
  ))

  css_mobile <- NULL
  if (!is.null(mobile)) {

    if (!("areas" %in% names(mobile))) {
      stop(paste("grid_template() mobile argument must contain list with `areas` definition.",
                 "See documentation for examples."))
    }

    css_grid_template_areas <- data_frame_to_css_grid_template_areas(mobile$areas)
    css_mobile <- shiny::tags$style(paste(
      "@media screen and (max-width: 768px) {",
        "#{{ grid_id }} {",
          grid_container_css(css_grid_template_areas, mobile$rows_height, mobile$cols_width),
        "}",
      "}"
    ))
  }

  template <- shiny::tagList(
    css_default,
    css_mobile,
    shiny::tags$div(id = "{{ grid_id }}", area_tags)
  ) %>% htmltools::renderTags()

  return(list(template = template$html, area_names = area_names))
}

#' Use CSS grid template in Shiny UI
#'
#' @param grid_template grid template created with grid_template() function
#' @param id id of grid
#' @param container_style character - string of custom CSS for the main grid container
#' @param area_styles list of custom CSS styles for provided area names
#' @param display_mode replaces areas HTML content with <area name> text. Used by display_grid() function
#' @param ... areas HTML content provided by named arguments
#'
#' @return Rendered HTML ready to use by Shiny UI. See \code{htmltools::htmlTemplate()} for more details.
#'
#' @details
#' Grids can be nested.
#'
#' @importFrom stats setNames
#'
#' @examples
#' myGrid <- grid_template(default = list(
#'   areas = rbind(
#'     c("header", "header", "header"),
#'     c("menu",   "main",   "right1"),
#'     c("menu",   "main",   "right2")
#'   ),
#'   rows_height = c("50px", "auto", "100px"),
#'   cols_width = c("100px", "2fr", "1fr")
#' ))
#'
#' subGrid <- grid_template(default = list(
#'   areas = rbind(
#'     c("top_left", "top_right"),
#'     c("bottom_left", "bottom_right")
#'   ),
#'   rows_height = c("50%", "50%"),
#'   cols_width = c("50%", "50%")
#' ))
#'
#' if (interactive()){
#' library(shiny)
#' library(shiny.semantic)
#' shinyApp(
#'   ui = semanticPage(
#'     grid(myGrid,
#'          container_style = "border: 1px solid #f00",
#'          area_styles = list(header = "background: #0099f9",
#'                             menu = "border-right: 1px solid #0099f9"),
#'          header = div(shiny::tags$h1("Hello CSS Grid!")),
#'          menu = checkbox_input("example", "Check me", is_marked = FALSE),
#'          main = grid(subGrid,
#'                      top_left = calendar("my_calendar"),
#'                      top_right = div("hello 1"),
#'                      bottom_left = div("hello 2"),
#'                      bottom_right = div("hello 3")
#'          ),
#'          right1 = div(
#'            toggle("toggle", "let's toggle"),
#'            multiple_checkbox("mycheckbox", "mycheckbox",
#'                              c("option A","option B","option C"))),
#'          right2 = div("right 2")
#'     )
#'   ),
#'   server = shinyServer(function(input, output) {})
#' )
#' }
#' @export
grid <- function(grid_template, id = paste(sample(letters, 5), collapse = ''),
                 container_style = "", area_styles = list(), display_mode = FALSE, ...) {

  if (display_mode) {
    # For debugging mode just display area name
    template_values <- as.list(setNames(grid_template$area_names, grid_template$area_names))
  } else {
    template_values <- list(...)
  }

  template_values$grid_id <- id
  template_values$custom_style_grid_container <- container_style

  for (name in grid_template$area_names) {
    key <- glue::glue("{name}_custom_css")
    template_values[[key]] <- ifelse(name %in% names(area_styles), area_styles[[name]], "")
  }

  do.call(function(...) htmltools::htmlTemplate(text_ = grid_template$template, ...), template_values)
}

#' Display grid template in a browser for easy debugging
#'
#' @param grid_template generated by grid_template() function
#' @param page_function function to use to create a page, for example semanticPage
#' @details
#' Opens a browser and displays grid template with styled border to highlight existing areas.
#'
#' Warning: CSS can't be displayed in RStudio viewer pane. CSS grid is supported only by modern browsers.
#' You can see list of supported browsers here: https://www.w3schools.com/css/css_grid.asp
#' @importFrom stats setNames
#' @importFrom grDevices rainbow
#' @export
display_grid <- function(grid_template, page_function = shiny::bootstrapPage) {

  # Apply style for debugging
  n <- length(grid_template$area_names)
  styles <- lapply(
    rainbow(n), # Pick rainbow colors and then change opacity to 44
    function(color) glue::glue(
      "border: 1px dotted #888;",
      "font-size: 2em;",
      "padding: 20px;",
      "background: {substr(color, 1, 7)}44"
    )
  ) %>% unlist
  area_styles <- as.list(setNames(styles, grid_template$area_names))

  shiny::runApp(list(
      ui = page_function(
        grid(
          grid_template,
          container_style = "border: 1px dashed #000",
          area_styles = area_styles,
          display_mode = TRUE
        )
      ),
      server = function(input, output) {}
    ),
    launch.browser = TRUE
  )

}
